cfg/builder/builder_finalize.cc

```diff
@@ -130,92 +131,102 @@ void CFGBuilder::sanityCheck(core::Context ctx, CFG &cfg) {
+LocalRef maybeDealias(core::Context ctx, CFG &cfg, LocalRef what, UnorderedMap<LocalRef, LocalRef> &aliases) {
+    if (what.isSyntheticTemporary(ctx, cfg)) {
+        auto fnd = aliases.find(what.id());
         if (fnd != aliases.end()) {
             return fnd->second;
-        } else {
-            return what;
         }
-    } else {
-        return what;
     }
+    return what;
 }
 
 /**
  * Remove aliases from CFG. Why does this need a separate pass?
  * because `a.foo(a = "2", if (...) a = true; else a = null; end)`
  */
 void CFGBuilder::dealias(core::Context ctx, CFG &cfg) {
-    vector<UnorderedMap<core::LocalVariable, core::LocalVariable>> outAliases;
-
+    vector<UnorderedMap<LocalRef, LocalRef>> outAliases;
     outAliases.resize(cfg.maxBasicBlockId);
+
     for (auto it = cfg.forwardsTopoSort.rbegin(); it != cfg.forwardsTopoSort.rend(); ++it) {
         auto &bb = *it;
         if (bb == cfg.deadBlock()) {
             continue;
         }
-        UnorderedMap<core::LocalVariable, core::LocalVariable> &current = outAliases[bb->id];
+        auto &current = outAliases[bb->id];
+
         if (!bb->backEdges.empty()) {
+            // Take the intersection of all of the back edges' aliases.
             current = outAliases[bb->backEdges[0]->id];
-        }
-
-        for (BasicBlock *parent : bb->backEdges) {
-            UnorderedMap<core::LocalVariable, core::LocalVariable> other = outAliases[parent->id];
-            for (auto it = current.begin(); it != current.end(); /* nothing */) {
-                auto &el = *it;
-                auto fnd = other.find(el.first);
-                if (fnd != other.end()) {
-                    if (fnd->second != el.second) {
-                        current.erase(it++);
+            for (auto i = 1; i < bb->backEdges.size(); i++) {
+                auto *parent = bb->backEdges[i];
+                const auto &other = outAliases[parent->id];
+                for (auto it = current.begin(); it != current.end(); /* nothing */) {
+                    auto &el = *it;
+                    auto fnd = other.find(el.first);
+                    if (fnd != other.end()) {
+                        if (fnd->second != el.second) {
+                            current.erase(it++);
+                        } else {
+                            ++it;
+                        }
                     } else {
-                        ++it;
+                        current.erase(
+                            it++); // note: this is correct but too conservative. In particular for loop headers
                     }
-                } else {
-                    current.erase(it++); // note: this is correct but to conservative. In particular for loop headers
                 }
             }
         }
 
+        // Overapproximation of the set of variables that have aliases.
+        // Will have false positives, but no false negatives. Avoids an expensive inner loop below.
```

Comment:
This optimization avoids an inner loop below that, otherwise, makes the loop `O(# bindings * # of aliases)`. On a file I was benchmarking, no invalidations ever occurred, so it was wasting time on that loop.

I made it an overapproximation because otherwise this needs to be a map with reference counting which seems like more effort than is needed.